fs/shm/shmfs.c

```diff
@@ -0,0 +1,370 @@
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+
+#include <sys/mman.h>
+
+#include <nuttx/fs/ioctl.h>
+#include <nuttx/fs/shmfs.h>
+
+#if defined (CONFIG_BUILD_KERNEL)
+#include <nuttx/arch.h>
+#include <nuttx/pgalloc.h>
+#include <nuttx/sched.h>
+#include <nuttx/mm/vm_map.h>
+#endif
+
+#include "shmfs_private.h"
+#include "inode/inode.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#ifndef min
+#define min(a,b) ((a) < (b) ? (a) : (b))
+#endif
+
+#ifndef max
+#define max(a,b) ((a) > (b) ? (a) : (b))
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int shmfs_close(FAR struct file *filep);
+static ssize_t shmfs_read(FAR struct file *filep, FAR char *buffer,
+                          size_t buflen);
+static ssize_t shmfs_write(FAR struct file *filep, FAR const char *buffer,
+                           size_t buflen);
+static int shmfs_truncate(FAR struct file *filep, off_t length);
+
+#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
+static int shmfs_unlink(FAR struct inode *inode);
+#endif
+
+static FAR void *shmfs_mmap(FAR struct file *filep, off_t start,
+                            size_t length);
+
+static int shmfs_munmap(FAR struct task_group_s *group,
+                        FAR struct inode *inode, void *vaddr, size_t length);
+
+/****************************************************************************
+ * Public Data
+ ****************************************************************************/
+
+const struct file_operations shmfs_operations =
+{
+  NULL,             /* open */
+  shmfs_close,      /* close */
+  shmfs_read,       /* read */
+  shmfs_write,      /* write */
+  NULL,             /* seek */
+  NULL,             /* ioctl */
+  shmfs_truncate,   /* truncate */
+  shmfs_mmap,       /* mmap */
+  shmfs_munmap,     /* munmap */
+  NULL              /* poll */
```

Comment:
Appending is a bit harder if done this way. It is not illegal to have a stray , in the end of an initializer field. I would do it like this for convenience:
```

NULL, /* poll */
#ifndef CONFIG_DISABLE_PSEUDOFS_OPERATIONS
shmfs_unlink,    /* unlink */
#endif
foo, /* whatever new fancy field comes next */
```

Reply:
Truncate creates this problem; it should probably be common in mountpt_operations. Or then have a separate "ftruncate" for file_operations?

Reply:
Of course, sorry misunderstood :)